core/src/main/java/com/datastax/oss/driver/internal/core/config/typesafe/DefaultProgrammaticConfigBuilder.java

```diff
@@ -0,0 +1,225 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.config.typesafe;
+
+import com.datastax.oss.driver.api.core.config.DriverConfigLoader;
+import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
+import com.datastax.oss.driver.api.core.config.DriverOption;
+import com.datastax.oss.driver.api.core.config.ProgrammaticConfigBuilder;
+import com.datastax.oss.driver.shaded.guava.common.base.Preconditions;
+import com.datastax.oss.protocol.internal.util.collection.NullAllowingImmutableMap;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigFactory;
+import com.typesafe.config.ConfigValueFactory;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import net.jcip.annotations.NotThreadSafe;
+
+@NotThreadSafe
+public class DefaultProgrammaticConfigBuilder implements ProgrammaticConfigBuilder {
+
+  public static final Supplier<Config> DEFAULT_FALLBACK_SUPPLIER = ConfigFactory::defaultReference;
+
+  private final NullAllowingImmutableMap.Builder<String, Object> values =
+      NullAllowingImmutableMap.builder();
+  private final Supplier<Config> fallbackSupplier;
+  private final String rootPath;
+
+  private String currentProfileName = DriverExecutionProfile.DEFAULT_NAME;
+
+  /**
+   * @param fallbackSupplier the supplier that will provide fallback configuration for options that
+   *     haven't been specified programmatically.
+   * @param rootPath the root path used in non-programmatic sources (fallback reference.conf and
+   *     system properties).
+   */
+  public DefaultProgrammaticConfigBuilder(Supplier<Config> fallbackSupplier, String rootPath) {
+    this.fallbackSupplier = fallbackSupplier;
+    this.rootPath = rootPath;
+  }
+
+  public DefaultProgrammaticConfigBuilder() {
+    this(DEFAULT_FALLBACK_SUPPLIER, "datastax-java-driver");
+  }
+
+  private ProgrammaticConfigBuilder with(@NonNull DriverOption option, @Nullable Object value) {
+    String path = option.getPath();
+    if (!DriverExecutionProfile.DEFAULT_NAME.equals(currentProfileName)) {
+      path = "profiles." + currentProfileName + "." + path;
+    }
+    if (!rootPath.isEmpty()) {
+      path = rootPath + "." + path;
+    }
+    return with(path, value);
+  }
+
+  private ProgrammaticConfigBuilder with(@NonNull String path, @Nullable Object value) {
+    values.put(path, value);
+    return this;
+  }
+
+  @NonNull
+  @Override
+  public ProgrammaticConfigBuilder startProfile(@NonNull String profileName) {
+    Preconditions.checkNotNull(profileName);
```

Comment:
1. Nit: couldn't we switch from Guava to regular Java and use `Objects.requireNonNull` instead?
2. My IDE complains about the return value of these two methods not being assigned, so I'd suggest the following idiom instead:

```
currentProfileName = Objects.requireNonNull(profileName);
```